activerecord/lib/active_record/connection_adapters.rb

```diff
@@ -13,19 +13,25 @@ class << self
       #
       # == Example
       #
-      #   ActiveRecord::ConnectionAdapters.register("megadb", "MegaDB::ActiveRecordAdapter", "mega_db/active_record_adapter")
-      #
       #   ActiveRecord::ConnectionAdapters.register("mysql", "ActiveRecord::ConnectionAdapters::TrilogyAdapter", "active_record/connection_adapters/trilogy_adapter")
       #
+      # Use the +:active_record_register_adapters+ hook to register custom adapters before database configs are loaded.
+      #
+      # == Example
+      #
+      #   ActiveSupport.on_load(:active_record_register_adapters) do
```

Comment:
Do we really need a callback for this? Maybe I missed some discussion?

Reply:
I hope not! Let me walk you through how I got here.

Adapters need to be registered _after Active Record is loaded_ since registration is called on `ActiveRecord::ConnectionAdapters`, but _before the connection config is loaded and validated_ from within the `ConnectionHandler`. That comes from the `initialize_database` railtie here:
https://github.com/rails/rails/blob/438cad462638b02210fc48b700c29dcd0428a8b7/activerecord/lib/active_record/railtie.rb#L305-L311

So you can only register adapters after AR is loaded, but you can't use `on_load(:active_record` because the above railtie is run by the framework earlier than any of the same hook that can be registered in the application, and because there is no `prepend:` type sequencing on these hooks, the one that validates the connections will always run first because it is declared first.

This hook would then fire at exactly the right time and in the right context for other adapters to be registered.

The only alternative I can see is to further delay validating the adapter in the connection, but that means we would end up with connection handler pools we know to be invalid, that we are hoping an adapter will be registered for later. I guess we can move back to using an `after_initialize` railtie to explicitly validate them. But I'm not sure that's better.

Reply:
> Adapters need to be registered after Active Record is loaded

While this sounds literally true (the `ActiveRecord` constant needs to be there), I don't think it's actually true in the `on_load` sense. I'm not immediately seeing why anything more than `require "active_record"` would be needed.

Reply:
Oh boy, I'm overcomplicating things.

Yes, that appears to work, so long as the app has `require "active_record"` the registration can be called directly in an initializer and called early but not too early. I've updated the integration test to reflect that.